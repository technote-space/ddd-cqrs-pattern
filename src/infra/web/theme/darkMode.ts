import type { Dispatch, Reducer } from '$/web/shared/store';
import type { IDarkModeContext, IDarkMode, StoreContext } from '$/web/theme/darkMode';
import { useSelector } from 'react-redux';
import { singleton } from 'tsyringe';

@singleton()
export class DarkModeContext implements IDarkModeContext {
  private readonly TOGGLE_DARK_MODE = 'TOGGLE_DARK_MODE';

  public getKey(): string {
    return 'darkMode';
  }

  public getInitialState(): StoreContext {
    return { darkMode: true };
  }

  public getReducerMapObject(): Record<string, Reducer<StoreContext>> {
    return {
      [this.TOGGLE_DARK_MODE]: (store) => ({
        ...store,
        darkMode: !store?.darkMode,
      }),
    };
  }

  public persistTargets(): Array<keyof StoreContext> {
    return ['darkMode'];
  }

  public useDarkMode(): boolean {
    return useSelector((state: { setting: StoreContext }) => state.setting.darkMode);
  }

  public toggleDarkMode(dispatch: Dispatch) {
    dispatch({ type: this.TOGGLE_DARK_MODE });
  }
}

@singleton()
export class DarkMode implements IDarkMode {
  // eslint-disable-next-line
  public useColorModeValue<T>(lightModeValue: T, darkModeValue: T): T {
    return lightModeValue;
  }

  public toggleDarkMode(): () => void {
    return () => {
      //
    };
  }
}
