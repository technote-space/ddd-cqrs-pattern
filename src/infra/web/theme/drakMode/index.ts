import type { useDarkMode as useDarkModeType, useToggleDarkMode as useToggleDarkModeType } from '$/web/theme/darkMode';
import { container } from 'tsyringe';

/* istanbul ignore next */
export const useDarkMode: useDarkModeType = () => container.resolve<useDarkModeType>('useDarkMode')();
/* istanbul ignore next */
export const useToggleDarkMode: useToggleDarkModeType = () => container.resolve<useToggleDarkModeType>('useToggleDarkMode')();
