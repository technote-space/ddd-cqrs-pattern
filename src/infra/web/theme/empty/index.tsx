import type { ITheme, Props } from '$/web/theme';
import type { VFC } from 'react';
import { memo } from 'react';
import { singleton } from 'tsyringe';
import { BaseComponent } from '@/web/shared/component';

@singleton()
export class EmptyTheme extends BaseComponent<Props> implements ITheme {
  public constructor() {
    super();
  }

  protected getComponent(): VFC<Props> {
    const component = memo(({ children }) => <div className="empty-theme">
      {children}
    </div>);
    component.displayName = 'EmptyTheme';

    return component;
  }
}
