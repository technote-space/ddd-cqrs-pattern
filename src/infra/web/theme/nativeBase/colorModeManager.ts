import type { IContext } from '$/web/shared/store';
import type { Reducer } from '$/web/shared/store';
import type { ColorMode, StorageManager } from 'native-base';
import { useMemo } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { singleton } from 'tsyringe';

type StoreContext = {
  colorMode: ColorMode;
};

export type IColorModeManagerContext = IContext<StoreContext>;

@singleton()
export class ColorModeContext implements IColorModeManagerContext {
  public static readonly SET_COLOR_MODE = 'SET_COLOR_MODE';

  public getKey(): string {
    return 'colorMode';
  }

  public getInitialState(): StoreContext {
    return { colorMode: 'dark' };
  }

  public getReducerMapObject(): Record<string, Reducer<StoreContext>> {
    return {
      [ColorModeContext.SET_COLOR_MODE]: (store, action) => ({
        ...store,
        colorMode: action.colorMode,
      }),
    };
  }

  public persistTargets(): Array<keyof StoreContext> {
    return ['colorMode'];
  }
}

export const useStorageManager = (): StorageManager => {
  const dispatch = useDispatch();
  /* istanbul ignore next */
  const selector = (state: { colorMode: StoreContext }): ColorMode => state.colorMode.colorMode ?? 'dark';
  const colorMode = useSelector(selector);

  return useMemo(() => ({
    get: async () => {
      return colorMode;
    },
    set: (colorMode: ColorMode) => {
      dispatch({ type: ColorModeContext.SET_COLOR_MODE, colorMode });
    },
  }), [dispatch, colorMode]);
};
