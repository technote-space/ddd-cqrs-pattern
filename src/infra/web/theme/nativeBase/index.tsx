import type { IContextProvider } from '$/web/shared/contextProvider';
import type { VFC } from 'react';
import type { PropsWithChildren } from 'react';
import { NativeBaseProvider, extendTheme, Box } from 'native-base';
import { memo } from 'react';
import { singleton } from 'tsyringe';

@singleton()
export class NativeBaseThemeProvider implements IContextProvider {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  private readonly __provider: VFC<PropsWithChildren<any>>;

  public constructor() {
    const theme = extendTheme({
      colors: {
        // https://smart-swatch.netlify.app/
        // primary: {
        //   50: '#E3F2F9',
        //   100: '#C5E4F3',
        //   200: '#A2D4EC',
        //   300: '#7AC1E4',
        //   400: '#47A9DA',
        //   500: '#0088CC',
        //   600: '#007AB8',
        //   700: '#006BA1',
        //   800: '#005885',
        //   900: '#003F5E',
        // },
      },
      config: {
        useSystemColorMode: true,
      },
    });

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const provider = memo(({ children }: PropsWithChildren<any>) => {
      return <NativeBaseProvider theme={theme}>
        <Box
          display="flex"
          height="100vh"
          overflow="auto"
          _light={{ bg: '#fffaf0' }}
          _dark={{ bg: '#1a202c' }}
        >
          {children}
        </Box>
      </NativeBaseProvider>;
    });
    provider.displayName = 'NativeBaseThemeProvider';
    this.__provider = provider;
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  public getProvider(): VFC<PropsWithChildren<any>> {
    return this.__provider;
  }
}
