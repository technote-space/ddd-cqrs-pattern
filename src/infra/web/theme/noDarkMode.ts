import type { IDarkMode } from '$/web/theme/darkMode';

export class NoDarkMode implements IDarkMode {
  // eslint-disable-next-line
  public useColorModeValue<T>(lightModeValue: T, darkModeValue: T): T {
    return lightModeValue;
  }

  public toggleColorMode(): () => void {
    return () => {
      //
    };
  }
}
