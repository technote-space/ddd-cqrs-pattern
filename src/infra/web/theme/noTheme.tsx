import type { ITheme } from '$/web/theme';
import type { VFC, ReactElement, ReactNode } from 'react';
import { memo } from 'react';
import { singleton } from 'tsyringe';

@singleton()
export class NoTheme implements ITheme {
  private readonly __component: VFC<{ children?: ReactNode }>;

  public constructor() {
    this.__component = NoTheme.getComponent();
  }

  public render(children?: ReactNode): ReactElement {
    const Component = this.__component;
    return <Component>
      {children}
    </Component>;
  }

  private static getComponent(): VFC {
    const component = memo(({ children }) => <>
      {children}
    </>);
    component.displayName = 'ThemeProvider';

    return component;
  }
}
