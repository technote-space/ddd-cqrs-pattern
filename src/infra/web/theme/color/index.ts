import type { useColor as useColorType, ThemeColor } from '$/web/theme/color';
import { container } from 'tsyringe';
import { useDarkMode } from '@/web/theme/drakMode';

export const useColor: useColorType = (key: keyof ThemeColor) => {
  const themeColor = container.resolve<ThemeColor>('ThemeColor');
  const darkMode = useDarkMode();

  return darkMode ? themeColor[key].dark : themeColor[key].light;
};
