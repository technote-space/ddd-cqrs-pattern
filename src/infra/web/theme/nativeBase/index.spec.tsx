import renderer from 'react-test-renderer';
import { container } from 'tsyringe';
import { NativeBaseThemeProvider } from '@/web/theme/nativeBase';

describe('NativeBaseTheme', () => {
  it('should return theme', () => {
    container.registerInstance('ThemeColor', {
      background: { dark: '#1a202c' },
      text: { dark: '#fff' },
    });
    const theme = new NativeBaseThemeProvider();
    const Provider = theme.getProvider();
    const tree = renderer.create(<Provider>
      <div>test</div>
    </Provider>);
    expect(tree).toMatchSnapshot();
  });
});
