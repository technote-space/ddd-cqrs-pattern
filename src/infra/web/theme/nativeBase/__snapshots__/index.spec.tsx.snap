// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NativeBaseTheme should return theme 1`] = `
<div
  className="css-view-1dbjc4n"
  style={
    Object {
      "WebkitFlex": 1,
      "flex": 1,
      "msFlex": "1 1 0%",
    }
  }
>
  <div
    className="css-view-1dbjc4n"
    style={
      Object {
        "backgroundColor": "rgba(255,250,240,1.00)",
        "display": "flex",
        "height": "100vh",
        "overflowX": "auto",
        "overflowY": "auto",
      }
    }
  >
    <div>
      test
    </div>
  </div>
</div>
`;
